pkg/util/ciutil/detect.go

```diff
@@ -19,85 +19,128 @@ import (
-// CI. See https://github.com/watson/ci-info/blob/master/index.js
-var detectors = map[System]detector{
-	AppVeyor:                    envVarDetector{envVar: []string{"APPVEYOR"}},
-	AWSCodeBuild:                envVarDetector{envVar: []string{"CODEBUILD_BUILD_ARN"}},
-	AtlassianBamboo:             envVarDetector{envVar: []string{"bamboo_planKey"}},
-	AtlassianBitbucketPipelines: envVarDetector{envVar: []string{"BITBUCKET_COMMIT"}},
-	AzurePipelines:              envVarDetector{envVar: []string{"TF_BUILD"}},
-	Buildkite:                   envVarDetector{envVar: []string{"BUILDKITE"}},
-	CircleCI:                    envVarDetector{envVar: []string{"CIRCLECI"}},
-	Codeship:                    envValueDetector{envMap: map[string]string{"CI_NAME": "codeship"}},
-	Drone:                       envVarDetector{envVar: []string{"DRONE"}},
-	GenericCI:                   envVarDetector{envVar: []string{"GENERIC_CI_SYSTEM"}},
-	GitHub:                      envVarDetector{envVar: []string{"GITHUB_WORKFLOW"}},
-	GitLab:                      envVarDetector{envVar: []string{"GITLAB_CI"}},
-	GoCD:                        envVarDetector{envVar: []string{"GO_PIPELINE_LABEL"}},
-	Hudson:                      envVarDetector{envVar: []string{"HUDSON_URL"}},
-	Jenkins:                     envVarDetector{envVar: []string{"JENKINS_URL", "BUILD_ID"}},
-	MagnumCI:                    envVarDetector{envVar: []string{"MAGNUM"}},
-	Semaphore:                   envVarDetector{envVar: []string{"SEMAPHORE"}},
-	TaskCluster:                 envVarDetector{envVar: []string{"TASK_ID", "RUN_ID"}},
-	TeamCity:                    envVarDetector{envVar: []string{"TEAMCITY_VERSION"}},
-	Travis:                      envVarDetector{envVar: []string{"TRAVIS"}},
-}
-
-// detector detects whether we're running in a particular CI system.
-type detector interface {
-	// IsCI returns true if we are currently running in this particular CI system.
-	IsCI() bool
-}
-
-// envVarDetector is a detector that uses the existence of a set of environment variables to determine if the system
-// is a CI system. All variaibles must be present.
-type envVarDetector struct {
-	envVar []string
-}
-
-// IsCI returns true if any of the detector's associated environment variables are set.
-func (c envVarDetector) IsCI() bool {
-	for _, e := range c.envVar {
-		if os.Getenv(e) == "" {
-			return false
-		}
-	}
-	return true
-}
+// CI. See https://github.com/watson/ci-info/blob/master/index.js.
+// For any CI system for which we detect additional env vars, the type of `System` is that is
+// specific to that CI system. The rest, even though we detect if it is that CI system, may not have an
+// implementation that detects all useful env vars, and hence just uses the `baseCI` struct type.
+var detectors = map[SystemName]system{
+	AppVeyor: baseCI{
+		Name:            AppVeyor,
+		EnvVarsToDetect: []string{"APPVEYOR"},
+	},
+	AWSCodeBuild: baseCI{
```

Comment:
For any system that just uses `baseCI`, it essentially means we _don't_ support it, right? We can detect that `pulumi` is running on `AWSCodeBuild`, but we don't have any logic to populate the expected metadata vars, right? So the only CI/CD system "name" is set, but other than that, the SHA, BuildURL, etc. will not be?

Reply:
That's correct. I added comment at the top of this map specifying that exact thing.

Reply:
Also, to be clear the reason why the other var properties won't end up being set in the environment bag is, because of https://github.com/pulumi/pulumi/blob/master/cmd/util.go#L507. We only set the value in the environment bag if the value is not an empty string.